src/providers/hana/qgshanaconnection.cpp

```diff
@@ -0,0 +1,791 @@
+                              "SCHEMA_NAME IN (%2) AND %3) "
+                              "INNER JOIN "
+                              "(SELECT TABLE_OID AS TABLE_OID_2, COMMENTS AS TABLE_COMMENTS FROM SYS.TABLES WHERE IS_USER_DEFINED_TYPE = 'FALSE') "
+                              "ON TABLE_OID = TABLE_OID_2" );
+
+  const QString sqlViews = QStringLiteral(
+                             "SELECT SCHEMA_NAME, VIEW_NAME, COLUMN_NAME, DATA_TYPE_NAME, VIEW_COMMENTS FROM "
+                             "(SELECT * FROM SYS.VIEW_COLUMNS WHERE "
+                             "VIEW_OID IN (SELECT OBJECT_OID FROM OWNERSHIP WHERE OBJECT_TYPE = 'VIEW' AND %1) AND "
+                             "SCHEMA_NAME IN (%2) AND %3) "
+                             "INNER JOIN "
+                             "(SELECT VIEW_OID AS VIEW_OID_2, COMMENTS AS VIEW_COMMENTS FROM SYS.VIEWS) "
+                             "ON VIEW_OID = VIEW_OID_2" );
+
+
+  QMultiHash<QPair<QString, QString>, QgsHanaLayerProperty> layers;
+
+  auto addLayers = [&]( const QString & sql, bool isView )
+  {
+    PreparedStatementRef stmt = mConnection->prepareStatement( QgsHanaUtils::toUtf16( sql ) );
+    stmt->setNString( 1, NString( schema.isEmpty() ? u"%" : schema.toStdU16String() ) );
+    QgsHanaResultSetRef rsLayers = QgsHanaResultSet::create( stmt );
+
+    while ( rsLayers->next() )
+    {
+      QgsHanaLayerProperty layer;
+      layer.schemaName = rsLayers->getString( 1 );
+      layer.tableName = rsLayers->getString( 2 );
+      QString geomColumnType = rsLayers->getString( 4 );
+      bool isGeometryColumn = ( geomColumnType == "ST_GEOMETRY" || geomColumnType == "ST_POINT" );
+      layer.geometryColName = isGeometryColumn ? rsLayers->getString( 3 ) : "";
+      layer.tableComment = rsLayers->getString( 5 );
+      layer.isView = isView;
+
+      QPair<QString, QString> layerKey( layer.schemaName, layer.tableName );
+      if ( allowGeometrylessTables )
+      {
+        int layersCount = layers.count( layerKey );
+        if ( !isGeometryColumn && layersCount >= 1 )
+          continue;
+        if ( layersCount == 1 )
+        {
+          QgsHanaLayerProperty firstLayer = layers.values( layerKey )[0];
+          if ( firstLayer.geometryColName.isEmpty() )
+          {
+            if ( isGeometryColumn )
+              layers.remove( layerKey );
+            else
+              continue;
+          }
+        }
+      }
+
+      layers.insert( layerKey, layer );
+    }
+    rsLayers->close();
+  };
+
+  try
+  {
+    QString sql = sqlTables.arg( sqlOwnerFilter, sqlSchemaFilter, sqlDataTypeFilter );
+    addLayers( sql, false );
+
+    sql = sqlViews.arg( sqlOwnerFilter, sqlSchemaFilter, sqlDataTypeFilter );
+    addLayers( sql, true );
+  }
+  catch ( const Exception &ex )
+  {
+    throw QgsHanaException( ex.what() );
+  }
+
+  QVector<QgsHanaLayerProperty> list;
+  for ( QPair<QString, QString> key : layers.uniqueKeys() )
```

Comment:
`const QPair<QString, QString>&` won't require a copy.

Reply:
Done in 2267454.